Add optimization summary documents (#14999)
eclipse-openj9/openj9: doc/compiler/optimizer/GlobalOptimizationsSummary.md

on the stack. Synchronizations on non-escaping objects can also be eliminated. Some
library methods are known to not cause the object to escape; these methods are special
cased in the analysis.

Stack allocations can be done in two forms: contiguous or non-contiguous.
- Non-contiguous allocations represent the best case scenario from a performance viewpoint
as the object and each of its constituent fields are represented by distinct temps.
This enables each of the fields to be considered by other analyses (e.g. dead store etc.)
completely independently of the others. Note that this can only be done if the allocation
is not used as an object (e.g. passed out into a call which dereferences the object).

- Contiguous allocations mean that the object is allocated on the stack but the fields
are all part of the object on the stack. References to the fields still have to use
the underlying object (except that it's on the stack instead of the heap).

An early pass that splits (duplicates) one allocation into multiple allocations along
different control flow paths (allocation splitting) is used to maximize the opportunities

> for escape analysis.

Reviewer: I agree with @hzongaro There probably is an escape analysis talk and/or presentation from him that we could link to if it is in the open. 

Other points that we probably should mention : `There is a notion of "heapifying" a stack allocated object in cases even when an escape point is detected for the object but the escape point is a cold block. In such cases it is worth keeping the object on stack on the hot path and only if the cold path is taken, then the object is allocated on the heap and the contents of the object on the stack are copied into the newly allocated heap object. As part of heapification, all references from autos/parms to the stack allocated object are updated to point to the heapified object instead. This means that it is safe to execute any code where the object escapes from that point on.`

> for escape analysis.

Reviewer 2: EA overview -> https://youtu.be/JMpqwLVwmk8

> for escape analysis.

Reviewer: Other potential additions:

`Our escape analysis implementation also has a notion of de-memoization/re-memoization. This pertains to some special classes in the JCL: Integer, Long, Short, Byte etc. that memoize some instances of those classes in a static array that they maintain (only for some ranges of values). These classes would use these arrays to return an already allocated/memoized object rather than allocate a new object every time (if possible) when `valueOf` is called. These arrays therefore also become escape points for objects that are allocated and stored into them. De-memoization attempts to get around this by converting valueOf calls to new allocations before escape analysis is attempted on the assumption that allocating such objects on the stack would be even more preferable to memoizing them in the arrays. If stack allocation was not possible due to some reason, the de-memoized allocations are re-memoized, leaving the code as it was originally and an attempt is also made to inline the re-memoized valueOf since the inlining of these calls is suppressed in the inliner to give de-memoization a chance later.`

> for escape analysis.

Reviewer: Finally, on the Power platform (maybe AArch64 should use it too if it has a weak memory model, attn : @0xdaryl ) there is a "flow sensitive escape analysis" pass that gets run. 

`Flow sensitive escape analysis tries to find places in the code where an object "has not escaped yet" rather than the notion of "does the object escape anywhere" that is the hallmark of the flow insensitive escape analysis that runs on non-Power platforms. The reason for doing this flow sensitive analysis is because we want to optimize away the lwsync instructions that must be emitted after allocations to ensure that other threads see a properly initialized object. Flow sensitive escape analysis attempts to move such lysyncs together (i.e. combine) as much as possible. This includes trying to move an lwsync for an object to another lwsync (either for the same object (after a constructor if the object has final fields) or a different object), a monexit or volatile store (that do lwsyncs for different reasons) but all only on the basis that the object has not escaped yet.`

## Explicit New Initialization
Explicit New Initialization attempts to skip zero initialization of objects where possible.
Basically it scans forward in the basic block from the point of the allocation and
tries to find explicit definitions of fields (say in constructors) that occur before
any use of the field. In these cases the zero initialization of fields (required according
to Java spec, so no field is uninitialized) can be skipped. Note that for reference

> fields the presence of a GC point before any user-specified definition means that

Reviewer: The notion of what qualifies as a GC point in the JIT typically depends on the query `node->canGCandReturn()`

> fields the presence of a GC point before any user-specified definition means that

Author: There is a hyperlink link to node->canGCandReturn() in a later section `Live Variables for GC (local and global)`. I'll move that link to the first time when GC point is mentioned. 

```
## Live Variables for GC (local and global)
Live Variables for GC attempts to minimize the number of locals that need to be zero
initialized upon method entry. The basic idea is to perform liveness analysis and find
out if any locals are live at any [GC point](https://github.com/eclipse/omr/blob/81b79405da6c7c960e611a8b2b12fd5861543330/compiler/il/OMRNode.cpp#L2298)
```

the field must be zero initialized. Peeking inside calls is done to maximize the
number of explicit field initializations seen by the analysis. In addition, for
some special array allocations (e.g. char arrays that are used by `String`, often do
not need zero initialization as they are filled in by `arraycopy` calls immediately),
we can skip zero initialization completely.

## Global Copy Propagation (Dead Store Elimination, Dead Structure Removal)
These three analyses use use/def info to perform the traditional optimizations. The

> analysis is only done for locals (statics/fields omitted). Dead structure removal

Reviewer 3: ```suggestion
analysis is only done for locals (statics/fields are omitted). Dead structure removal
```

basically tries to find a region of code with no side-effects and having no definition
that is used outside the structure; in such cases the (dead) region of code can be
eliminated.

## Redundant Monitor Elimination
Redundant Monitor Elimination uses value numbering to perform nested monitor removal

> (e.g. syncenter on the same object twice followed by two syncexits). Use value

Reviewer 3: ```suggestion
(e.g. syncenter on the same object twice followed by two syncexits). It uses value
```

number information and structure to perform lock coarsening (e.g. syncenter/syncexit
on some object followed by another syncenter/syncexit on the same object). In this
case the extent of code across which the lock is held is increased and one syncexit/syncenter
can be removed. Deadlock avoidance is ensured by examining the code where the lock
would be held after coarsening and making sure there are no calls or lock/unlock
operations in that region of code. Sometimes additional syncenters/syncexits may be
added on ‘cold’ paths to avoid some syncenters/syncexits on ‘hot’ paths.

## Global Register Allocator (GRA)
GRA is a tactical solution for achieving the effect of global register assignment,

> even though the underlying code generators have local register allocators. Basic idea

Reviewer: This tactical solution has survived now for over 20 years, i.e. I'd suggest it has earned the right to no longer be referred to as tactical :) @0xdaryl

is that global register dependencies are introduced in the IL trees; these dependencies
(when evaluated by the code generators) ensure that a certain value is in a certain
register on all control flow paths into a basic block. Stores/loads from memory are
changed into stores/loads from registers in the IL (register numbers specify which
physical register the value is placed into). Global register candidates are chosen
by various optimizer passes like PRE, induction variable detection etc. and appropriate
priorities are given to the candidates based on the number of references and live

> range (number of blocks). Advantages are that most of the code is platform neutral

Reviewer 3: ```suggestion
range (number of blocks). The advantages are that most of the code is platform neutral
```

Reviewer: Suggestion : `most of the code is platform neutral and is` -> `most of the code is platform neutral apart from a few simple evaluators that must be added to a new platform code generator, thus drastically reducing the development cost of adding a powerful global register allocator. It is also `

> and it is relatively cheap in compile time in comparison to (expected cost of) a

Reviewer 3: ```suggestion
and it is relatively cheap in compile time in comparison to (the expected cost of) a
```

graph-coloring allocator.

## Live Variables for GC (local and global)
Live Variables for GC attempts to minimize the number of locals that need to be zero

> initialized upon method entry. Basic idea is to perform liveness analysis and find

Reviewer 3: ```suggestion
initialized upon method entry. The basic idea is to perform liveness analysis and find
```

out if any locals are live at any [GC point](https://github.com/eclipse/omr/blob/81b79405da6c7c960e611a8b2b12fd5861543330/compiler/il/OMRNode.cpp#L2298)
that is not dominated by (user defined) stores on all control flow paths. If a local
is uninitialized at a GC point, explicit zero initialization must be inserted at the method entry.

> Note that this situation is fairly rare (only occurs in some uncommon cases involving jsr).

Reviewer 3: "Involving JSR": do you mean JSR292? JSR by itself just refers to Java Specification Request.

Reviewer: No, a long long time ago in a galaxy far far away, there used to be a bytecode known as `jsr` (jump to sub routine) that was used to implement throw/catch/finally when Java was first released. Because this code essentially caused some abnormal control flow, it could make it appear as though a variable was uninitialized on some code path. The `jsr` bytecode was removed a long time back. 

I believe the main situation in which such zero initializations may be needed in today's JIT compiler may be around some optimization creating such an appearance of uninitialized variables. 

I wonder if we should add the word "reference"  to "local" since the GC cares only about reference locals.

Another function of this analysis is to minimize the number of object stack slots that
are examined at a GC point (locals that are not live at a GC point need not be examined at all).

Local Live Variables for GC does not perform a full blown liveness analysis; instead
it performs a forward walk through the CFG till the first GC point is seen and collects
all the locals that have been stored till that point and marks the other locals as
requiring zero initialization.

Global Live Variables for GC collects zero initialization information as explained above;
in addition it also populates each basic block with information about which locals
are live on entry to the block. This is then used by the code generator to only

> (selectively) switch on bits in the GC stack maps for those reference locals that

Reviewer: There is some historical documentation that @0xdaryl had found related to GC stack maps. Maybe he can comment on whether there was some plan to move to the open in the future.

> (selectively) switch on bits in the GC stack maps for those reference locals that

Reviewer 2: I believe the material Vijay mentions could be used as a starting point for a GC map overview but more will need to be added to it to make it a useful contribution.  This is on our rather long list of concepts we wish to document as part of the tech debt initiative, but no document exists just yet to link to in this document..

the inlining of these calls is suppressed in the inliner to give de-memoization a chance later.

Allocation sinking is an optimization that helps Escape Analysis. It basically moves the `new` node
as close as possible to the corresponding `<init>` call. These two trees can be separated by
arbitrary control flow on occasions when the Java code is like `x = new X(foo());` and so `foo()`
is between the `new` and the `<init>` call. If `foo()` gets inlined, then one can even get non
trivial control flow between the `new` and the `<init>` call.

On PowerPC platform, there is a "Flow Sensitive Escape Analysis" pass.
Flow Sensitive Escape Analysis tries to find places in the code where an object
"has not escaped yet" rather than the notion of "does the object escape anywhere" that is
the hallmark of the Flow Insensitive Escape Analysis that runs on non-Power platforms.
The reason for doing this Flow Sensitive Escape Analysis is because we want to optimize away the
`lwsync` instructions that must be emitted after allocations to ensure that other threads
see a properly initialized object. Flow Sensitive Escape Analysis attempts to move such

> `lysyncs` together (i.e. combine) as much as possible. This includes trying to move an `lwsync`

Reviewer: typo : `lysync` -> `lwsync`
